app/experimenter/nimbus-ui/src/hooks/useCommonForm.tsx

```diff
@@ -0,0 +1,160 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+import React, { useState } from "react";
+import { RegisterOptions, FieldError } from "react-hook-form";
+import Form from "react-bootstrap/Form";
+import { useForm } from "react-hook-form";
+
+// exported for testing
+export const overviewFieldNames = [
+  "name",
+  "hypothesis",
+  "application",
+  "publicDescription",
+] as const;
+
+export const audienceFieldNames = [
+  "channel",
+  "firefoxMinVersion",
+  "targetingConfigSlug",
+  "populationPercent",
+  "totalEnrolledClients",
+  "proposedEnrollment",
+  "proposedDuration",
+] as const;
+
+export const metricsFieldNames = [
+  "primaryProbeSetIds",
+  "secondaryProbeSetIds",
+] as const;
+
+type OverviewFieldNames = typeof overviewFieldNames[number];
+type AudienceFieldNames = typeof audienceFieldNames[number];
+type MetricsFieldNames = typeof metricsFieldNames[number];
+
+type FieldNames = OverviewFieldNames | AudienceFieldNames | MetricsFieldNames;
+
+// TODO: 'any' type on `onChange={(selectedOptions) => ...`,
+// it wants this, but can't seem to coerce it into SelectOption type
+// type SelectedOption = {
+//   value: ValueType<any, true>;
+//   action: ActionMeta<any>;
+// };
+type SelectOption = { label: string; value: string };
+
+export function useCommonForm(
+  defaultValues: Record<string, any>,
+  isServerValid: boolean,
+  submitErrors: Record<string, string[]>,
+  setSubmitErrors: React.Dispatch<React.SetStateAction<Record<string, any>>>,
+) {
+  const {
+    handleSubmit,
+    register,
+    reset,
+    errors,
+    formState,
+    getValues,
+  } = useForm({
+    mode: "onTouched",
+    defaultValues,
+  });
+
+  // We must compare the last onFocus value to the last onBlur value rather than
+  // checking `dirtyFields[fieldName]` because the field stays dirty if the value
+  // differs from the default and form errors are displayed.
+  const [previousValue, setPreviousValue] = useState<string | null>(null);
```

Comment:
Hmm, we wouldn't even need this if we wanted to clear the errors `onChange` of inputs/selects rather than check `onBlur`/`onFocus`. I think that would be better UX anyway since client-side validation updates on change.

Reply:
I updated this to clear `onChange` for consistency/simplification. Can revert easily but once it's confirmed we want this `onChange` instead (awaiting response in Slack) I'll squash the 2nd commit/make sure the commit message and tests reflect "onChange" rather than "onBlur".

Reply:
Is this that thing that we were concerned about performance issues with?

Reply:
@jodyheavener Yes, I remember us bringing up that `onBlur` would probably be more performant, but this ended up being a lot more simplistic than I thought since it's just an `if` check on if a key/value exists in a small object. Sure it calls the function/check `onChange` but I feel that's fine.

Reply:
Sounds good, this explanation makes sense and I trust this should be fine. We can always review later if something feels off.